google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/sdk/CloudSdkServiceManager.java

```diff
@@ -43,6 +74,176 @@ public void onNewCloudSdkServiceTypeSelected(CloudSdkServiceType newServiceType)
+        (sdkService, status) -> {
+          switch (status) {
+            case READY:
+            case INVALID:
+            case NOT_AVAILABLE:
+              installationCompletionLatch.countDown();
+              break;
+            case INSTALLING:
+              // continue waiting for completion.
+              break;
+          }
+        };
+
+    if (installInProgress) {
+      cloudSdkService.addStatusUpdateListener(sdkStatusUpdateListener);
+      callback.log(GctBundle.getString("appengine.deployment.status.preparing.sdk") + "\n");
+    } else {
+      // no need to wait for install if unsupported or completed.
+      installationCompletionLatch.countDown();
+    }
+
+    CountDownLatch blockingCompletedLatch = new CountDownLatch(1);
+
+    // wait for SDK to be ready and trigger the actual deployment if it properly installs.
+    ProgressManager.getInstance()
+        .run(
+            new Task.Backgroundable(project, progressMessage, true) {
+              boolean userCancelled;
+
+              @Override
+              public void run(@NotNull ProgressIndicator indicator) {
+                try {
+                  while (installationCompletionLatch.getCount() > 0) {
+                    // wait interruptibility to check for user cancel each second.
+                    installationCompletionLatch.await(1, SECONDS);
+                    if (checkIfCancelled()) {
+                      userCancelled = true;
+                      break;
+                    }
+                  }
+
+                  // at this point the installation should be either ready or user cancelled.
+                  ApplicationManager.getApplication()
+                      .invokeLater(() -> doRun(cloudSdkService, runnable, callback, userCancelled));
+
+                } catch (InterruptedException e) {
+                  /* valid cancellation exception, no handling needed. */
+                } finally {
+                  // remove the notification listener regardless of waiting outcome.
+                  cloudSdkService.removeStatusUpdateListener(sdkStatusUpdateListener);
+                  // unblock the waiting latch for external point of synchronization
+                  blockingCompletedLatch.countDown();
+                }
+              }
+            });
+
+    return blockingCompletedLatch;
+  }
+
+  @VisibleForTesting
+  boolean checkIfCancelled() {
+    return ProgressManager.getInstance().getProgressIndicator().isCanceled();
+  }
+
+  private void doRun(
+      CloudSdkService cloudSdkService,
+      @Nullable Runnable runnable,
+      CloudSdkPreconditionCheckCallback callback,
+      boolean userCancelled) {
+    // check the status of SDK after install.
+    SdkStatus postInstallSdkStatus = cloudSdkService.getStatus();
+    switch (postInstallSdkStatus) {
+      case INSTALLING:
+        String message =
+            GctBundle.message(
+                userCancelled
+                    ? "appengine.deployment.error.cancelled"
```

Comment:
why is there still deployment specific stuff in here if its now meant to be generic (i.e. also handling local run)?

Reply:
Good point, we need another message as a parameter for user cancel message then. I'll update.